src/reducers.rs

```diff
@@ -385,66 +265,186 @@ impl Reducer for Script {
-        self.set_seed(seed.clone());
-        Ok(Box::new(()))
+        if self.child.is_none() {
+            self.spawn_child()?;
+        }
+
+        let response = self.request(Request::New(NewRequest { seed: seed.path().into() }))?;
+        match response {
+            Response::New(NewResponse { state }) => Ok(Box::new(state)),
+            otherwise => {
+                let program = self.program.to_string_lossy().to_string();
+                self.misbehaving_reducer_script(format!(
+                    "Expected a `Response::New` in response to a `Request::New` request; \
+                     got `{:?}` from '{}'",
+                    otherwise,
+                    program
+                ))
+            }
+        }
     }
 
-    fn clone_state(&self, _: &Box<Any + Send>) -> Box<Any + Send> {
-        Box::new(())
+    fn clone_state(&self, state: &Box<Any + Send>) -> Box<Any + Send> {
+        Box::new(Self::downcast(state).clone())
     }
 
     fn next_state(
         &mut self,
-        _: &test_case::Interesting,
-        _: &Box<Any + Send>
+        seed: &test_case::Interesting,
+        state: &Box<Any + Send>
     ) -> error::Result<Option<Box<Any + Send>>> {
-        Ok(Some(Box::new(())))
+        // It's possible that we killed the child for misbehaving since we
+        // generated this state, so we can't assert that the child exists.
+        if self.child.is_none() {
+            self.spawn_child()?;
+        }
+
+        let state = Self::downcast(state);
+        let response = self.request(Request::Next(NextRequest {
+            seed: seed.path().into(),
+            state: state.clone(),
+        }))?;
+
+        match response {
+            Response::Next(NextResponse { next_state }) => {
+                Ok(next_state.map(|ns| Box::new(ns) as Box<Any + Send>))
+            }
+            otherwise => {
+                let program = self.program.to_string_lossy().to_string();
+                self.misbehaving_reducer_script(format!(
+                    "Expected a `Response::Next` in response to a `Request::Next` request; \
+                     got `{:?}` from '{}'",
+                    otherwise,
+                    program
+                ))
+            }
+        }
     }
 
     fn next_state_on_interesting(
         &mut self,
         new_seed: &test_case::Interesting,
-        _old_seed: &test_case::Interesting,
-        _prev_state: &Box<Any + Send>
+        old_seed: &test_case::Interesting,
+        state: &Box<Any + Send>
+    ) -> error::Result<Option<Box<Any + Send>>> {
+        if self.child.is_none() {
+            self.spawn_child()?;
+        }
+
+        let state = Self::downcast(state);
+        let response = self.request(Request::NextOnInteresting(NextOnInterestingRequest {
+            new_seed: new_seed.path().into(),
+            old_seed: old_seed.path().into(),
+            state: state.clone(),
```

Comment:
Hm, I sort of wish this passed the state that generated the valid reduction, instead of the text of the reduction. The inverse, where files are passed around seems overly complex to work with for some reducers.

It's basically impossible to go text -> state (Need to keep a hashmap? That would waste a lot of memory. Assumming 1MB test file, and 10000 attempted states, that's roughly 10 gigabytes of file data), while the reducers can easily generate the text from the interesting state using the existing reduce function.

Reply:
> Hm, I sort of wish this passed the state that generated the valid reduction, instead of the text of the reduction.

That's what `state` is, so you actually get both ;)

The `Reducer` actor is keeping track of (and garbage collecting) which state generated which potential reduction for you.

Reply:
When using `preduce_reducer_script::Reducer`, you aren't multiplexing different states anymore. The `self` *is* the state, and the runner is handling multiplexing and serialization for you.

Reply:
I've ported `chunks.py` to `chunks.rs`; let me push it in a second here.